packages/medusa/src/utils/transaction/transaction-builder.ts

```diff
@@ -0,0 +1,358 @@
+import { TransactionStepsDefinition } from "./types"
+
+export type ActionHandler = {
+  [type: string]: (data: any, context: any) => Promise<any>
+}
+
+interface InternalStep extends TransactionStepsDefinition {
+  next?: InternalStep | InternalStep[]
+  depth: number
+  parent?: InternalStep | null
+}
+
+export class TransactionBuilder {
+  private steps: InternalStep
+
+  constructor(steps?: TransactionStepsDefinition) {
+    this.load(steps)
+  }
+
+  load(steps?: TransactionStepsDefinition) {
+    this.steps = {
+      depth: -1,
+      parent: null,
+      next: steps
+        ? JSON.parse(
+            JSON.stringify((steps.action ? steps : steps.next) as InternalStep)
+          )
+        : undefined,
+    }
+
+    this.updateDepths(this.steps, {}, 1, -1)
+    return this
+  }
+
+  addAction(
+    action: string,
+    handler: ActionHandler,
```

Comment:
question: shouldn't we have an handler for both invoke and compensation unless there is no compensation?

Reply:
at the moment no. The handlers are not being used yet.
but it is very likely that these handlers will belong to the class managing Workflows. So I will remove them from here